### PR TITLE
Made BlocksProjectiles upgradable

### PR DESCRIPTION
**Pull request by reaperrr**
Needed for TS firestorm defense and laser fences.

### CODE REVIEW

File: OpenRA.Mods.Common/Effects/Bullet.cs [hunk 1]
```diff
@@ -164,8 +165,11 @@ public void Tick(World world)
 			if (info.ContrailLength > 0)
 				contrail.Update(pos);
 
-			if (ticks++ >= length || (info.Blockable && world.ActorMap
-				.GetUnitsAt(world.Map.CellContaining(pos)).Any(a => a.HasTrait<IBlocksProjectiles>())))
+			if (ticks++ >= length
+				|| (info.Blockable && world.ActorMap
+					.GetUnitsAt(world.Map.CellContaining(pos))
+					.Any(a => a.TraitsImplementing<BlocksProjectiles>()
+					.Any(b => !b.IsTraitDisabled))))
```
Comment (atlimit8) [inline OpenRA.Mods.Common/Effects/Bullet.cs]: Please add a level of indentation to show that it is inside the previous Any.

File: OpenRA.Mods.Common/Effects/Missile.cs [hunk 2]
```diff
@@ -201,7 +201,10 @@ public void Tick(World world)
 			var shouldExplode = (pos.Z < 0) // Hit the ground
 				|| (dist.LengthSquared < info.CloseEnough.LengthSquared) // Within range
 				|| (info.RangeLimit != 0 && ticks > info.RangeLimit) // Ran out of fuel
-				|| (info.Blockable && world.ActorMap.GetUnitsAt(cell).Any(a => a.HasTrait<IBlocksProjectiles>())) // Hit a wall or other blocking obstacle
+				|| (info.Blockable && world.ActorMap
+					.GetUnitsAt(world.Map.CellContaining(pos))
+					.Any(a => a.TraitsImplementing<BlocksProjectiles>()
+					.Any(b => !b.IsTraitDisabled))) // Hit a wall or other blocking obstacle
```
Comment (atlimit8) [inline OpenRA.Mods.Common/Effects/Missile.cs]: Please add a level of indentation to show that it is inside the previous Any as also requested above.

**Comment by reaperrr** [comment]
Updated.

File: OpenRA.Mods.Common/Effects/Bullet.cs [hunk 3]
```diff
@@ -164,8 +165,11 @@ public void Tick(World world)
 			if (info.ContrailLength > 0)
 				contrail.Update(pos);
 
-			if (ticks++ >= length || (info.Blockable && world.ActorMap
-				.GetUnitsAt(world.Map.CellContaining(pos)).Any(a => a.HasTrait<IBlocksProjectiles>())))
+			if (ticks++ >= length
```
Comment (Unit158) [inline OpenRA.Mods.Common/Effects/Bullet.cs]: You'll have to check two times per tick now. The current location and the next location.

Comment (reaperrr) [inline OpenRA.Mods.Common/Effects/Bullet.cs]: Why and how?

Comment (Unit158) [inline OpenRA.Mods.Common/Effects/Bullet.cs]: it actually shouldn't matter in practice due to order latency, so never mind :)

File: OpenRA.Mods.Common/Effects/Bullet.cs [hunk 4]
```diff
@@ -164,8 +165,11 @@ public void Tick(World world)
...
+				|| (info.Blockable && world.ActorMap
+					.GetUnitsAt(world.Map.CellContaining(pos))
+					.Any(a => a.TraitsImplementing<BlocksProjectiles>()
+						.Any(b => !b.IsTraitDisabled))))
```
Comment (pchote) [inline OpenRA.Mods.Common/Effects/Bullet.cs]: This is a bit of a monster to have inside a conditional check, so could you please pull this out into an `IsBlockedByActor()` method?

File: OpenRA.Mods.Common/Effects/Missile.cs [hunk 5]
```diff
@@ -201,7 +201,10 @@ public void Tick(World world)
 			var shouldExplode = (pos.Z < 0) // Hit the ground
 				|| (dist.LengthSquared < info.CloseEnough.LengthSquared) // Within range
 				|| (info.RangeLimit != 0 && ticks > info.RangeLimit) // Ran out of fuel
-				|| (info.Blockable && world.ActorMap.GetUnitsAt(cell).Any(a => a.HasTrait<IBlocksProjectiles>())) // Hit a wall or other blocking obstacle
+				|| (info.Blockable && world.ActorMap
```
Comment (pchote) [inline OpenRA.Mods.Common/Effects/Missile.cs]: Likewise here

**Comment by pchote** [comment]
Needs a rebase, sorry.

**Comment by reaperrr** [comment]
@atlimit8 Is there or will there be something to replace

`.Any(a => a.TraitsImplementing<BlocksProjectiles>().Any(b => !b.IsTraitDisabled))`

with an Info equivalent?

**Comment by atlimit8** [comment]
If you mean check if the actor type has a `BlocksProjectiles` trait that is enabled by default, then yes; however, that (being enabled/disabled) is part of the traits state (may change).

In an upcoming PR I plan on allowing something like:
```
.WhereType(projectileBlockerTypes).Any(a => a.TraitsImplementing<BlocksProjectiles>().Any(Exts.IsTraitEnabled))
```
where `projectileBlockerTypes` is a cached `HashSet<uint>` of `ActorTypeID`s : 
```
projectileBlockerTypes = self.World.Map.Rules.ActorsWhereTraitInfo<BlocksProjectilesInfo>();
```

That would allow for filtering out all non-blockers with an amortized cost of O(N) where N is the number of actors so the cost (O(N * log T) where N is the number of possible blockers and T is the number of trait pairs in the dictionary) of the
```
.Any(a => a.TraitsImplementing<BlocksProjectiles>().Any(Exts.IsTraitEnabled))
```
will only be applied for actors that might block (usually 0, otherwise generally 1).

Basically, that should change the cost from O(N * log T) to an amortized O(N + B * log T) where N is the number of actors, B is the number of possible blockers and T is the number of trait pairs in the dictionary. 


I am working on a branch that will add `WhereOwner*` filters to replace common delegates and then basing the aforementioned branch on that using comment feedback to iron out naming style.

**Comment by atlimit8** [comment]
In case you missed it in my previous post:
You should replace `b => !b.IsTraitDisabled` with `Exts.IsTraitEnabled`

**Comment by reaperrr** [comment]
> You should replace `b => !b.IsTraitDisabled` with `Exts.IsTraitEnabled`

Done.

**Comment by RoosterDragon** [comment]
Is the duplicated ```BlockedByActor``` method something we can move into one place, maybe on ```BlocksProjectiles```?

**Comment by abcdefg30** [comment]
Looks good to me. :+1:

**Comment by penev92** [comment]
Maybe have the `if` (both of them) as
```
if (ticks++ >= length || (info.Blockable && AnyBlockersAt(world, pos)))
```

and move `BlockedByActor()` to `BlocksProjectiles` as
`public static bool AnyBlockersAt(World world, WPos pos)` ?

**Comment by reaperrr** [comment]
Rebased and updated.

**Comment by penev92** [comment]
Works :+1:

**Comment by penev92** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/8e31fb4bbbfbb7ef7a202a821fca4d3dc02f25e6...1f39ddebb0b97deb8495003e4983a93379f9ef08)
